typing/typemod.ml

```diff
@@ -85,14 +87,64 @@ let extract_sig_open env loc mty =
 
 (* Compute the environment after opening a module *)
 
-let type_open_ ?used_slot ?toplevel ovf env loc lid =
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+
+let gen_mod_ident () =
+  let n = !mod_ident_counter in
+  incr mod_ident_counter;
+  Ident.create (Printf.sprintf "M#%d" n)
+
+let open_struct_level = ref 0
+let in_nested_struct () = !open_struct_level <> 0
+let enter_struct () = incr open_struct_level
+let leave_struct () = decr open_struct_level
+
+let open_generated s = String.contains s '#'
+
+let type_open_ ?used_slot ?toplevel ovf env loc me =
+  let report_illegal_mt mt loc env =
+    ignore (extract_sig_open env loc mt);
+    assert false
+  in
+  match me.pmod_desc with
+  | Pmod_ident lid ->
+      let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
+      begin
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
+      | Some env ->
+          let tme =
+            {
+              mod_desc=Tmod_ident (path, lid);
+              mod_loc=lid.loc;
+              mod_type=Mty_ident path;
+              mod_env=env;
+              mod_attributes=me.pmod_attributes
+            } in
+          None, tme, env
+      | None ->
+        let mt = (Env.find_module path env).md_type in
+        report_illegal_mt mt lid.loc env
+      end
+  | Pmod_structure _ | Pmod_apply _ | Pmod_constraint _ | Pmod_unpack _ |
+    Pmod_functor _ | Pmod_extension _ ->
+      enter_struct ();
+      let ident = gen_mod_ident () in
+      let tme = !type_module_fwd env me in
```

Comment:
type_module normally rejects Pmod_unpack within a functor body (unless the packaged module declares no type).  This is necessary for safety reason, and I'm concerned that this call to type_module_fwd does not pass the `funct_body` flag.

Reply:
Perhaps this is safe, because the types from the opened module cannot escape to the functor's body type itself.